tools/stack_decode.py

```diff
@@ -0,0 +1,80 @@
+#!/usr/bin/env python
+
+# Call addr2line as needed to resolve addresses in a stack trace,
+# de-interleaving the log lines from multiple threads if required.
+#
+# Two ways to call:
+#   1) No arguments means this script will treat stdin as log output from
+#   Envoy
+#   2) Some arguments means run a subprocess with that command line and send
+#   stderr through the script
+#
+# In each case this script will decode any backtrace log lines found and echo
+# back all non-Backtrace lines untouched.
+
+import collections
+import re
+import subprocess
+import sys
+
+# Process the log output looking for stacktrace snippets, print them out once
+# the entire stack trace has been read.  End when EOF received.
+def decode_stacktrace_log(input_source):
+  traces = {}
+  trace_begin_re = re.compile('^(.+)\[backtrace\] Backtrace obj<(.+)> thr<(\d+)')
+  stackaddr_re = re.compile('\[backtrace\] thr<(\d+)> #(\d+) (0x[0-9a-fA-F]+)$')
+  trace_end_re = re.compile('\[backtrace\] end backtrace thread (\d+)')
+
+  # build a dictionary indexed by thread_id:
+  #   value is a namedtuple (log_prefix, obj_file, [list of addresses])
+  Backtrace = collections.namedtuple('Backtrace', 'log_prefix obj_file, address_list')
+  try:
+    while True:
+      line = input_source.readline()
+      if line == '':
+        return # EOF
+      begin_trace_match = trace_begin_re.search(line)
+      if begin_trace_match:
+        log_prefix, objfile, thread_id = begin_trace_match.groups()
+        traces[thread_id] = Backtrace(log_prefix = log_prefix,
+                                      obj_file = objfile,
+                                      address_list = [])
+        continue
+      stackaddr_match = stackaddr_re.search(line)
+      if stackaddr_match:
+        thread_id, frame_idx, address = stackaddr_match.groups()
+        traces[thread_id].address_list.append((frame_idx, address))
+        continue
+      trace_end_match = trace_end_re.search(line)
+      if trace_end_match:
+        thread_id = trace_end_match.groups()[0]
+        output_stacktrace(thread_id, traces[thread_id])
+      else:
+        print line, # Pass through print all other log lines
+  except KeyboardInterrupt:
+    return
+
+# Output one stacktrace after passing it through addr2line with appropriate
+# options
+def output_stacktrace(thread_id, traceinfo):
+  log_prefix, object_file, stack = traceinfo
+  piped_input = ""
+  for stack_frame in stack:
+    piped_input += (stack_frame[1] + '\n')
```

Comment:
This index operation is actually picking the address out of the tuple (first element captured is the thread ID).  I'm going to change this slightly to have better clarity.